Update GitHub Actions and Set the Platform Requirement to macOS 11

PSchmiedmayer opened the pull request:
# Update GitHub Actions and Set the Platform Requirement to macOS 11

## :recycle: Current situation & Problem
Currently, GitHub Action workflows and other configuration files are located in every repo. This PR moves them to the Apodini/.github repo.
In addition, the GitHub Actions do not include macOS builds due to the lack of backward compatibility of the concurrency features introduced in Swift 5.5 to older Apple OS versions.

## :bulb: Proposed solution
The PR removes the local GitHub Action workflows and other configuration files and references the GitHub Action workflows and other configuration files in the [Apodini/.github](https://github.com/Apodini/.github) repo.
Xcode 13.2 introduced backward compatibility of, e.g., async/await to older Apple OS versions. The GitHub Actions in the [Apodini/.github](https://github.com/Apodini/.github) repo are updated to use the latest stable Swift versions on Linux and Xcode 13.2 on the GitHub macOS 11 build agents.
Also closes #397 and #396.

## 🛠 Current State
- [x] The PR is blocked by [bugs.swift.org: SR-15445 - swift test fails using Swift 5.5.2 on macOS 11: Library not loaded: @rpath/libswift_Concurrency.dylib](https://bugs.swift.org/browse/SR-15445) also documented on [forums.swift.org](https://forums.swift.org/t/swift-5-5-2-xcode-13-2-beta-fails-to-link-libswift-concurrency-dylib/53263). Shared with @DougGregor on [Twitter](https://twitter.com/PSchmiedmayer/status/1456559875471953920). Update: Has been fixed in https://github.com/apple/swift-package-manager/pull/3847, will probably end up in the Xcode 13.2.1.
- [x] The Swift NIO Async-Await Extensions are currently make with `@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)` which results in compile time errors on macOS 11. Swift NIO needs to update its `@available` markings so we can remove the `.unsafeFlags(["-Xfrontend", "-disable-availability-checking"])` Swift build flags in the Package.swift file. We can't tag a release with `unsafeFlags` in the Package.swift file. I have created a GitHub issue for the Swift NIO project: https://github.com/apple/swift-nio/issues/2000 and created a PR at https://github.com/apple/swift-nio/issues/2004. The PR is merged and the issue closed, wait until SwiftNIO tags a new release.

## :gear: Release Notes 
- Supports building for macOS 11

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/Apodini/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/Apodini/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/Apodini/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/Apodini/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer committed:
Update GitHub Actions

PSchmiedmayer committed:
Update GitHub Actions and Set the Platform Requirement to macOS 11

PSchmiedmayer committed:
Remove _Concurrency imports

PSchmiedmayer committed:
Try without xcodebuild

PSchmiedmayer committed:
Install additional dependencies

PSchmiedmayer committed:
Update Swift Packages

PSchmiedmayer committed:
Fix build error due to _NIOConcurrency

PSchmiedmayer committed:
Update GitHub build actions

PSchmiedmayer committed:
Update GitHub Actions

PSchmiedmayer committed:
Rename GitHub Actions

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update imports and fix SwiftLint error

PSchmiedmayer committed:
Add a Docker Test

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update build-and-test.yml

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update Dependencies

PSchmiedmayer committed:
Update GitHub Actions

lukaskollmer commented on .github/workflows/build-and-test.yml:
```diff
@@ -1,108 +1,105 @@
+  workflow_call:
+    inputs:
+      packagename:
+        description: 'Name of the package e.g., passed ass the scheme to xcodebuild on macOS builds'
```
```suggestion
        description: 'Name of the package e.g., passed as the scheme to xcodebuild on macOS builds'
```

PSchmiedmayer replied:
Thanks! 👍

PSchmiedmayer committed:
Update Docker Compose files

PSchmiedmayer committed:
Fix Type

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Fix SwiftLint warnings

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Improve GitHub Actions

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update build-and-test.yml

PSchmiedmayer committed:
Try new approach for testing the TestWebService

PSchmiedmayer committed:
Remove input

PSchmiedmayer committed:
Install dependencies for Linux builds

PSchmiedmayer committed:
Fix a few GitHub Action related issues

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update to latest Swift NIO release

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer commented:
@lukaskollmer @moritzsternemann Updating to the latest SwiftNIO version seems to break two separate features that I can reproduced locally and where I would like to have a second opinion before changing the test behavior in both cases.

As Xcode 13.2 doesn't yet deliver on its promise of enabling Swift Concurrency on macOS 11 (at least not reliably, see TestWebService on macOS) I am most probably going to remodel this PR just as a "We improved GitHub Actions by moving them to the .github" repo PR but before that it would be good to resolve the Swift NIO update related issues. Any input and hints from your side are welcome 🚀

## 1. Apodini Logger Tests
The Apodini Logger tests seems to have run into some tests assertions that seems to have changed:

1. https://github.com/Apodini/Apodini/blob/f2e102265ce5cc04089c35ead12e15a372deba98/Tests/ApodiniObserveTests/ApodiniLoggerTests.swift#L766
2. https://github.com/Apodini/Apodini/blob/f2e102265ce5cc04089c35ead12e15a372deba98/Tests/ApodiniObserveTests/ApodiniLoggerTests.swift#L769
3. https://github.com/Apodini/Apodini/blob/f2e102265ce5cc04089c35ead12e15a372deba98/Tests/ApodiniObserveTests/ApodiniLoggerTests.swift#L853
4. https://github.com/Apodini/Apodini/blob/f2e102265ce5cc04089c35ead12e15a372deba98/Tests/ApodiniObserveTests/ApodiniLoggerTests.swift#L856

They all seem to related to the fact that the "content-length" information with a value of 0 is no longer present. Would be interested to get your input at @moritzsternemann is currently working on ApodiniObserve and @lukaskollmer as a lot of NIO experience: Do you have any idea where this might come from and if we can safely assume that this is no significant regression and we just update our assumptions of what SwiftNIO provides? @philippzagar: You are also welcome to chime in if you have any good idea :)

## 2. Localhost Deployment Provider

There seems to be an issue with starting up the localhost web services for the localhost deployment prover tests: https://github.com/Apodini/Apodini/runs/4520198109?check_suite_focus=true#step:8:16924
```
/Users/runner/work/Apodini/Apodini/Tests/ApodiniDeployTests/LocalhostDeploymentProviderTests.swift:350: error: -[ApodiniDeployTests.LocalhostDeploymentProviderTests testLocalhostDeploymentProvider] : failed - Received error in HTTPRequestClientResponseDelegate: The operation couldn’t be completed. (AsyncHTTPClient.HTTPClient.NWPOSIXError error 1.)
/Users/runner/work/Apodini/Apodini/Tests/ApodiniDeployTests/LocalhostDeploymentProviderTests.swift:350: error: -[ApodiniDeployTests.LocalhostDeploymentProviderTests testLocalhostDeploymentProvider] : failed - Received error in HTTPRequestClientResponseDelegate: The operation couldn’t be completed. (AsyncHTTPClient.HTTPClientError error 1.)
/Users/runner/work/Apodini/Apodini/Tests/ApodiniDeployTests/LocalhostDeploymentProviderTests.swift:350: error: -[ApodiniDeployTests.LocalhostDeploymentProviderTests testLocalhostDeploymentProvider] : failed - Received error in HTTPRequestClientResponseDelegate: The operation couldn’t be completed. (AsyncHTTPClient.HTTPClientError error 1.)
/Users/runner/work/Apodini/Apodini/Tests/ApodiniDeployTests/LocalhostDeploymentProviderTests.swift:282: error: -[ApodiniDeployTests.LocalhostDeploymentProviderTests testLocalhostDeploymentProvider] : Asynchronous wait failed: Exceeded timeout of 15 seconds, with unfulfilled expectations: "Web Service response for /v1/ request", "Web Service response for /v1/greet/ request", "Web Service response for /v1/textMut/ request".
/Users/runner/work/Apodini/Apodini/Tests/ApodiniDeployTests/LocalhostDeploymentProviderTests.swift:313: error: -[ApodiniDeployTests.LocalhostDeploymentProviderTests testLocalhostDeploymentProvider] : Asynchronous wait failed: Exceeded timeout of 60 seconds, with unfulfilled expectations: "Did shut down servers", "ChildProcess did terminate".
```

@lukaskollmer Do you have a hint where this is coming from and how/if this is a regression based on SwiftNIO?

lukaskollmer commented:
@PSchmiedmayer i'll have more time to look into this tomorrow or on thursday, but my initial guess based on the error messages you commented would be that the AsyncHTTPClient (which is itself based on NIO) might be causing these errors

philippzagar commented:
@PSchmiedmayer @moritzsternemann I don't have time to have a closer look at it right now, but my first guess would just be that the new NIO version doesn't "collect" (or set) the content-length header anymore if there isn't any content (so if it is 0). I should be able to take a closer look on the weekend at the latest. Otherwise I'm sure that @moritzsternemann is an `ApodiniObserve` expert by now and will be able to quickly resolve these issues :)
Practically, I guess that you should be able to remove these assertions without any other changes, but please inspect the headers of the request as well as the metadata collected by `ApodiniObserve` just to be sure.

moritzsternemann commented:
I was looking through the changes in swift-nio from 2.33.0 to 2.36.0 but couldn't really find the related one. As I don't think the `content-length` header is required when it is zero, I think this is probably just reflected now in NIO and there is no harm in updating the tests to pass 👍

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update GitHub Actions

PSchmiedmayer committed:
Update ApodiniObsereve Tests

PSchmiedmayer committed:
Test docker image creation on push

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update Dockerfile

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Reduce number of Linux builds

PSchmiedmayer committed:
Add additional logs

PSchmiedmayer commented:
@moritzsternemann & @philippzagar Thanks for your input, I have updated to test cases to reflect your points and the current state of Swift NIO 👍

@lukaskollmer Also thanks to you for your input! I have further investigated the LocalHostDeploymentIssue. As far as I could investigate it, the HTTPClient is working as expected but the proxy web service is not starting and accepting requests as expected. I have added a few more logs in the tests logging more details about the HTTPClient errors. I also extended the timeout and manually send request to the proxy web services at e.g. `http://localhost:80/v1` using `curl` and received a `curl: (7) Failed to connect to localhost port 80: Connection refused` message. So my guess is that even though the logs indicate that the proxy web service is starting there seems to be an issue accepting requests. Do you have an idea at what I should further look at and where this might come from?

lukaskollmer commented:
@PSchmiedmayer I just had a look at this and I managed to fix it. Can I push onto this branch?

The issue seems to have been that ApodiniDeploy was unable to construct a HTTP request object for the AsyncHTTPClient for proxying a client request to the actual server (or rather the AsyncHTTPClient rejected ApodiniDeploy's request object) because the hostname and port were missing, because the HTTPRequest created by ApodiniNetworking to represent the client request contained a URL object where these fields were empty, which was the case because the HTTP request head (from which ApodiniNetworking constructs the URL) simply doesn't contain these kinds of information. So the change that caused this to become a crash was either in NIO changing the format of the request heads, or in the AsyncHTTPClient's changing its checking of url hostnames.

lukaskollmer committed:
Fix crash in ApodiniDeploy

PSchmiedmayer commented:
@lukaskollmer Thanks for taking a look at this! That's interesting, feel free to push to this branch 🚀💪

lukaskollmer commented:
@PSchmiedmayer done. let me know if this didn't fix things

PSchmiedmayer committed:
Update test and use hostname instead of bind address

PSchmiedmayer committed:
Fix SwiftLint errors

PSchmiedmayer commented:
@lukaskollmer Thanks a lot! I have adapted your changes a bit to use the hostname instead of the bind address and fixed a few assumptions here and there and all tests seem to pass now. There is still an issue with Swift Concurrency on macOS 11 but I think that Xcode 13.2.1 will fix this as indicated in its release notes, we will probably have to wait a bit until it arrives on the GitHub Actions runners but GitHub already seems to have started with adding it: https://github.com/actions/virtual-environments/pull/4765

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Actions and Tests

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Comment out testing the test web service on macOS

PSchmiedmayer committed:
Update Dependencies

PSchmiedmayer committed:
Merge branch 'develop' into feature/ciimprovements

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Actions with tags

PSchmiedmayer committed:
Update GitHub Actions with tags

PSchmiedmayer committed:
Update GitHub Action

PSchmiedmayer committed:
Update GitHub Actions

PSchmiedmayer committed:
Update TestWebService GitHub Action name to avoid conflics
